File: ios/App/App/Storage/HostTransactionJournal.swift
import Foundation

/// Secret-bearing transaction journal. Lives on the host-vault Keychain boundary.
/// Stores exact target registry / vault bytes so crash recovery is idempotent.
enum HostTransactionJournal {
    /// v2 captures exact legacy-source bytes so an interrupted explicit removal
    /// can resume without deleting credentials written after the operation began.
    /// v3 adds a metadata-only rename operation without changing registry shape.
    /// v4 allows one removal to delete additional dependent vault accounts
    /// (an environment's locally paired child records when their parent is
    /// removed) in the same crash-atomic operation.
    static let currentVersion = 4
    static let account = HostVault.journalAccount

    enum Kind: String, Codable, Sendable, Equatable {
        case add
        /// Adds one environment record without changing the selected host.
        case addEnvironment
        case switchSelected
        case rename
        case remove
    }

    enum Phase: String, Codable, Sendable, Equatable {
        /// Intent recorded; target bytes not yet applied (or apply interrupted).
        case intent
        /// Vault mutation applied (save or delete). Registry not yet confirmed.
        case vaultApplied
        /// Registry exact bytes applied. Journal may still exist until deleted.
        case registryApplied
    }

    struct Record: Codable, Sendable, Equatable {
        var version: Int
        var operationId: UInt64
        var kind: Kind
        var connectionId: ClientConnectionID
        var phase: Phase
        /// Exact bytes to land in `registry.json`. Recovery writes these, never re-encodes.
        var targetRegistryBytes: Data
        var targetVaultAccount: String?
        /// Exact vault payload. Recovery writes these bytes when present.
        var targetVaultBytes: Data?
        var deleteVaultAccount: String?
        /// Additional dependent vault accounts to delete with the primary one
        /// (v4). Never holds the primary `deleteVaultAccount`.
        var deleteVaultAccounts: [String]?
        /// Exact legacy source slots observed before an explicit imported-host removal.
        var legacySource: LegacyHostSourceSnapshot?
        /// Exact tombstone bytes, written only when every source slot was unchanged.
        var targetTombstoneBytes: Data?
    }

    enum Decode: Sendable, Equatable {
        case current(Record)
        case future
        case corrupt
    }

    /// Test-only crash/restart seams.
    enum Stage: Sendable, Equatable {
        case afterIntent
        case afterVaultApply
        case afterRegistryApply
    }

    static func encode(_ record: Record) throws -> Data {
        try HostRegistryCoding.encode(record)
    }

    static func decode(_ data: Data) -> Decode {
        do {
            guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let version = root["version"] as? Int
            else {
                return .corrupt
            }
            if version > currentVersion { return .future }
            if version == 1 {
                let legacy = try HostRegistryCoding.decode(LegacyRecordV1.self, from: data)
                return validated(
                    Record(
                        version: currentVersion,
                        operationId: legacy.operationId,
                        kind: legacy.kind,
                        connectionId: legacy.connectionId,
                        phase: legacy.phase,
                        targetRegistryBytes: legacy.targetRegistryBytes,
                        targetVaultAccount: legacy.targetVaultAccount,
                        targetVaultBytes: legacy.targetVaultBytes,
                        deleteVaultAccount: legacy.deleteVaultAccount,
                        deleteVaultAccounts: nil,
                        legacySource: nil,
                        targetTombstoneBytes: nil
                    )
                )
            }
            if version == 2 {
                let legacy = try HostRegistryCoding.decode(LegacyRecordV2.self, from: data)
                return validated(
                    Record(
                        version: currentVersion,
                        operationId: legacy.operationId,
                        kind: legacy.kind.current,
                        connectionId: legacy.connectionId,
                        phase: legacy.phase,
                        targetRegistryBytes: legacy.targetRegistryBytes,
                        targetVaultAccount: legacy.targetVaultAccount,
                        targetVaultBytes: legacy.targetVaultBytes,
                        deleteVaultAccount: legacy.deleteVaultAccount,
                        deleteVaultAccounts: nil,
                        legacySource: legacy.legacySource,
                        targetTombstoneBytes: legacy.targetTombstoneBytes
                    )
                )
            }
            if version == 3 {
                let legacy = try HostRegistryCoding.decode(LegacyRecordV3.self, from: data)
                return validated(
                    Record(
                        version: currentVersion,
                        operationId: legacy.operationId,
                        kind: legacy.kind,
                        connectionId: legacy.connectionId,
                        phase: legacy.phase,
                        targetRegistryBytes: legacy.targetRegistryBytes,
                        targetVaultAccount: legacy.targetVaultAccount,
                        targetVaultBytes: legacy.targetVaultBytes,
                        deleteVaultAccount: legacy.deleteVaultAccount,
                        deleteVaultAccounts: nil,
                        legacySource: legacy.legacySource,
                        targetTombstoneBytes: legacy.targetTombstoneBytes
                    )
                )
            }
            if version < currentVersion { return .corrupt }
            let record = try HostRegistryCoding.decode(Record.self, from: data)
            guard record.version == currentVersion else { return .corrupt }
            return validated(record)
        } catch {
            return .corrupt
        }
    }

    private static func validated(_ record: Record) -> Decode {
        guard record.version == currentVersion,
              let document = try? HostRegistryStore.migratedDocument(record.targetRegistryBytes)
        else { return .corrupt }
        let account = HostVault.account(for: record.connectionId)
        let dependentAccounts = record.deleteVaultAccounts ?? []
        guard Set(dependentAccounts).count == dependentAccounts.count,
              !dependentAccounts.contains(account),
              dependentAccounts.allSatisfy({ HostVault.connectionId(fromAccount: $0) != nil }),
              dependentAccounts.allSatisfy({ dependent in
                  HostVault.connectionId(fromAccount: dependent)
                      .map { document.host(id: $0) == nil } ?? false
              })
        else { return .corrupt }
        switch record.kind {
        case .add:
            guard record.targetVaultAccount == account,
                  record.targetVaultBytes?.isEmpty == false,
                  record.deleteVaultAccount == nil,
                  record.deleteVaultAccounts == nil,
                  record.legacySource == nil,
                  record.targetTombstoneBytes == nil,
                  document.host(id: record.connectionId) != nil,
                  document.selectedConnectionId == record.connectionId
            else { return .corrupt }
        case .addEnvironment:
            guard record.targetVaultAccount == account,
                  record.targetVaultBytes?.isEmpty == false,
                  record.deleteVaultAccount == nil,
                  record.deleteVaultAccounts == nil,
                  record.legacySource == nil,
                  record.targetTombstoneBytes == nil,
                  document.host(id: record.connectionId)?.environment != nil
            else { return .corrupt }
        case .switchSelected:
            guard record.targetVaultAccount == nil,
                  record.targetVaultBytes == nil,
                  record.deleteVaultAccount == nil,
                  record.deleteVaultAccounts == nil,
                  record.legacySource == nil,
                  record.targetTombstoneBytes == nil,
                  document.host(id: record.connectionId) != nil,
                  document.selectedConnectionId == record.connectionId
            else { return .corrupt }
        case .rename:
            guard record.targetVaultAccount == nil,
                  record.targetVaultBytes == nil,
                  record.deleteVaultAccount == nil,
                  record.deleteVaultAccounts == nil,
                  record.legacySource == nil,
                  record.targetTombstoneBytes == nil,
                  document.host(id: record.connectionId) != nil
            else { return .corrupt }
        case .remove:
            guard record.targetVaultAccount == nil,
                  record.targetVaultBytes == nil,
                  record.deleteVaultAccount == account,
                  (record.legacySource == nil) == (record.targetTombstoneBytes == nil),
                  document.host(id: record.connectionId) == nil
            else { return .corrupt }
        }
        return .current(record)
    }

    private struct LegacyRecordV1: Codable {
        var version: Int
        var operationId: UInt64
        var kind: Kind
        var connectionId: ClientConnectionID
        var phase: Phase
        var targetRegistryBytes: Data
        var targetVaultAccount: String?
        var targetVaultBytes: Data?
        var deleteVaultAccount: String?
        var clearLegacySource: Bool
    }

    private enum LegacyKindV2: String, Codable {
        case add
        case switchSelected
        case remove

        var current: Kind {
            switch self {
            case .add: .add
            case .switchSelected: .switchSelected
            case .remove: .remove
            }
        }
    }

    private struct LegacyRecordV2: Codable {
        var version: Int
        var operationId: UInt64
        var kind: LegacyKindV2
        var connectionId: ClientConnectionID
        var phase: Phase
        var targetRegistryBytes: Data
        var targetVaultAccount: String?
        var targetVaultBytes: Data?
        var deleteVaultAccount: String?
        var legacySource: LegacyHostSourceSnapshot?
        var targetTombstoneBytes: Data?
    }

    /// Version 3 shape: one deleted vault account per removal.
    private struct LegacyRecordV3: Codable {
        var version: Int
        var operationId: UInt64
        var kind: Kind
        var connectionId: ClientConnectionID
        var phase: Phase
        var targetRegistryBytes: Data
        var targetVaultAccount: String?
        var targetVaultBytes: Data?
        var deleteVaultAccount: String?
        var legacySource: LegacyHostSourceSnapshot?
        var targetTombstoneBytes: Data?
    }
}

struct HostJournalIO: Sendable {
    let vault: HostVault

    func load() throws -> HostTransactionJournal.Decode? {
        guard let data = try vault.load(account: HostTransactionJournal.account) else {
            return nil
        }
        return HostTransactionJournal.decode(data)
    }

    func loadRaw() throws -> Data? {
        try vault.load(account: HostTransactionJournal.account)
    }

    func save(_ record: HostTransactionJournal.Record) throws {
        let data = try HostTransactionJournal.encode(record)
        try vault.save(account: HostTransactionJournal.account, data: data)
    }

    func delete() throws {
        try vault.delete(account: HostTransactionJournal.account)
    }
}

File: android/app/src/main/kotlin/com/poracode/app/storage/HostTransactionJournal.kt
package com.poracode.app.storage

import com.poracode.app.model.ClientConnectionId
import com.poracode.app.model.RemoteJson
import java.util.Base64
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.int
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Secret-bearing, encrypted transaction journal with exact target bytes.
 *
 * Version 3 adds [Record.deleteVaultAccounts] so one atomic removal can retire
 * a parent record together with its dependent environment records and their
 * child credential vault accounts. Version 1/2 records stay readable: the new
 * list defaults to empty and [Record.deleteVaultAccount] keeps its exact
 * single-account meaning. A version bump here is a persisted-boundary change:
 * the pre-upgrade fixture in `HostTransactionJournalTest` pins that an old
 * journal still decodes and recovers.
 */
object HostTransactionJournal {
    const val VERSION = 3
    private const val OLDEST_SUPPORTED_VERSION = 1

    @Serializable
    enum class Kind { Add, Select, Remove, Rename }

    @Serializable
    enum class Phase { Intent, VaultApplied, RegistryApplied }

    @Serializable
    data class Record(
        val version: Int = VERSION,
        val operationId: Long,
        val kind: Kind,
        val connectionId: ClientConnectionId,
        val phase: Phase,
        val targetRegistryBase64: String,
        val targetVaultAccount: String? = null,
        val targetVaultBase64: String? = null,
        val deleteVaultAccount: String? = null,
        /**
         * Additional vault accounts retired by this transaction (dependent
         * environment child grants on a parent removal cascade). Version 3+.
         */
        val deleteVaultAccounts: List<String> = emptyList(),
        val clearLegacySource: Boolean = false,
        val importReceiptBase64: String? = null,
    ) {
        val targetRegistryBytes: ByteArray
            get() = Base64.getDecoder().decode(targetRegistryBase64)
        val targetVaultBytes: ByteArray?
            get() = targetVaultBase64?.let(Base64.getDecoder()::decode)

        fun withPhase(next: Phase): Record = copy(phase = next)
    }

    sealed class Decode {
        data class Current(val record: Record) : Decode()
        data object Future : Decode()
        data object Corrupt : Decode()
    }

    fun make(
        operationId: Long,
        kind: Kind,
        connectionId: ClientConnectionId,
        targetRegistryBytes: ByteArray,
        targetVaultAccount: String? = null,
        targetVaultBytes: ByteArray? = null,
        deleteVaultAccount: String? = null,
        deleteVaultAccounts: List<String> = emptyList(),
        clearLegacySource: Boolean = false,
        importReceiptBytes: ByteArray? = null,
    ): Record = Record(
        operationId = operationId,
        kind = kind,
        connectionId = connectionId,
        phase = Phase.Intent,
        targetRegistryBase64 = Base64.getEncoder().encodeToString(targetRegistryBytes),
        targetVaultAccount = targetVaultAccount,
        targetVaultBase64 = targetVaultBytes?.let(Base64.getEncoder()::encodeToString),
        deleteVaultAccount = deleteVaultAccount,
        deleteVaultAccounts = deleteVaultAccounts,
        clearLegacySource = clearLegacySource,
        importReceiptBase64 = importReceiptBytes?.let(Base64.getEncoder()::encodeToString),
    )

    fun encode(record: Record): ByteArray {
        val encoded = RemoteJson.parseToJsonElement(
            RemoteJson.encodeToString(record.copy(version = VERSION)),
        ).jsonObject
        val versioned = JsonObject(
            linkedMapOf("version" to JsonPrimitive(VERSION)) + encoded,
        )
        return RemoteJson.encodeToString(versioned).toByteArray(Charsets.UTF_8)
    }

    fun decode(bytes: ByteArray): Decode {
        val raw = bytes.toString(Charsets.UTF_8)
        val version = runCatching {
            RemoteJson.parseToJsonElement(raw).jsonObject["version"]?.jsonPrimitive?.int
        }.getOrNull() ?: return Decode.Corrupt
        if (version > VERSION) return Decode.Future
        if (version < OLDEST_SUPPORTED_VERSION) return Decode.Corrupt
        val record = runCatching {
            RemoteJson.decodeFromString<Record>(raw).requireValid()
        }.getOrNull()
            ?: return Decode.Corrupt
        return Decode.Current(record)
    }

    private fun Record.requireValid(): Record {
        require(operationId >= 0) { "Invalid journal operation" }
        require(targetRegistryBytes.isNotEmpty()) { "Missing target registry" }
        require((targetVaultAccount == null) == (targetVaultBase64 == null)) {
            "Incomplete target vault mutation"
        }
        require(targetVaultAccount == null || deleteVaultAccount == null) {
            "Conflicting vault mutation"
        }
        require(deleteVaultAccounts.size == deleteVaultAccounts.distinct().size) {
            "Duplicate deleted vault account"
        }
        require(deleteVaultAccounts.none { it == targetVaultAccount }) {
            "Conflicting cascade vault mutation"
        }
        require(deleteVaultAccounts.none { it == deleteVaultAccount }) {
            "Duplicate deleted vault account"
        }
        require(deleteVaultAccounts.all(::isVaultAccount)) {
            "Unexpected cascade vault account"
        }
        val expectedAccount = HostVault.account(connectionId)
        require(targetVaultAccount == null || targetVaultAccount == expectedAccount) {
            "Unexpected target vault account"
        }
        require(deleteVaultAccount == null || deleteVaultAccount == expectedAccount) {
            "Unexpected deleted vault account"
        }
        targetVaultBytes?.let { require(it.isNotEmpty()) { "Empty target vault payload" } }
        importReceiptBase64?.let { Base64.getDecoder().decode(it) }
        return this
    }

    private val VAULT_ACCOUNT_PATTERN = Regex("^host-vault\\.[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")

    private fun isVaultAccount(account: String): Boolean =
        VAULT_ACCOUNT_PATTERN.matches(account)
}
